# Data Wrangling

<h5>

**Wrangling** /ˈræŋ.ɡəl.ɪŋ/

the activity of taking care of, controlling, or moving animals, especially large animals such as cows or horses

</h5>

([Cambridge Dictionary](https://dictionary.cambridge.org/dictionary/english/wrangling))

![Cattle Wrangler - image from https://commons.wikimedia.org/wiki/File:Pioneer_Day_Wrangler.jpg](https://upload.wikimedia.org/wikipedia/commons/thumb/8/83/Pioneer_Day_Wrangler.jpg/320px-Pioneer_Day_Wrangler.jpg)

**[Data wrangling](https://en.wikipedia.org/wiki/Data_wrangling)** commonly refers to the transformation of data from one "input" format (e.g., `.csv` files from an experiment), to a different format (e.g., a tidy dataframe) that is more appropriate to the needs of an analysis. In the context of the ExPra experiments, you will use data wrangling techniques to implement the transformations and data cleaning steps specified in your preregistrations.

## Setup

### Setup Part 1: Install Packages

We will use [`tidyverse`](https://www.tidyverse.org/) packages to implement our data wrangling. The "tidyverse" is a series of packages which share a philosphy based around code and data structures that are (a) tidy, and (b) readable. You can install the tidyverse packages like so:

```
install.packages("tidyverse")
```

This includes many packages that we won't be using today, but which will be useful in other parts of the course (e.g., on Data Visualisation).

Remember, you should install packages in the console - never in a script that your share with others. This is because otherwise, your script will go to the effort of reinstalling a package *every time* it is run!

### Setup Part 2: Check the Packages Load

Now, we can test that the packages we will be using today actually load. You should be able to run this code without any errors:

In [23]:
options(repr.plot.width=3.5, repr.plot.height=3)

In [24]:
library(dplyr)
library(tidyr)

### Setup Part 3

Finally, check that you can access the dataset we'll be using in this session. The `starwars` dataset is a dataset built into R that contains details of characters from the Star Wars films:

In [25]:
print(starwars)

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Luke Sk~    172    77 blond      fair       blue            19   male  mascu~
 2 C-3PO       167    75 NA         gold       yellow         112   none  mascu~
 3 R2-D2        96    32 NA         white, bl~ red             33   none  mascu~
 4 Darth V~    202   136 none       white      yellow          41.9 male  mascu~
 5 Leia Or~    150    49 brown      light      brown           19   fema~ femin~
 6 Owen La~    178   120 brown, gr~ light      blue            52   male  mascu~
 7 Beru Wh~    165    75 brown      light      blue            47   fema~ femin~
 8 R5-D4        97    32 NA         white, red red             NA   none  mascu~
 9 Biggs D~    183    84 black      light      brown           24   male  mascu~
10 Obi-Wan~    182    77 auburn, w~ fair       blue-gray       57   male  mascu~
# i 77 m

This is snapshot shows an example of tidy data - a philosophy of organising data such that each observation (*character*) has a single row, with all variables tied to that character as a single column.

Now that we're all set up, let's dig into the data...

## Arrange

We can use the `arrange()` function to sort by variables in the dataframe. For example, we can arrange all characters in order of height (shortest to tallest) like so:

In [26]:
arrange(starwars, height)

name                  height mass hair_color skin_color         
1   Yoda                   66    17.0 white      green              
2   Ratts Tyerell          79    15.0 none       grey, blue         
3   Wicket Systri Warrick  88    20.0 brown      brown              
4   Dud Bolt               94    45.0 none       blue, grey         
5   R2-D2                  96    32.0 NA         white, blue        
6   R4-P17                 96      NA none       silver, red        
7   R5-D4                  97    32.0 NA         white, red         
8   Sebulba               112    40.0 none       grey, red          
9   Gasgano               122      NA none       white, blue        
10  Watto                 137      NA black      blue, grey         
11  Leia Organa           150    49.0 brown      light              
12  Mon Mothma            150      NA auburn     fair               
13  Cordé                 157      NA brown      light              
14  Nien Nunb             160    68.0 none       grey               
15  Shmi Skywalker        163      NA black      fair               
16  Ben Quadinaros        163    65.0 none       grey, green, yellow
17  Beru Whitesun lars    165    75.0 brown      light              
18  Dormé                 165      NA brown      light              
19  Padmé Amidala         165    45.0 brown      light              
20  Barriss Offee         166    50.0 black      yellow             
21  C-3PO                 167    75.0 NA         gold               
22  Jocasta Nu            167      NA white      fair               
23  Zam Wesell            168    55.0 blonde     fair, green, yellow
24  Wedge Antilles        170    77.0 brown      fair               
25  Palpatine             170    75.0 grey       pale               
26  Finis Valorum         170      NA blond      fair               
27  Luminara Unduli       170    56.2 black      yellow             
28  Eeth Koth             171      NA black      brown              
29  Luke Skywalker        172    77.0 blond      fair               
30  Greedo                173    74.0 NA         green              
... ...                   ...    ...  ...        ...                
58  Raymus Antilles       188     79  brown      light              
59  Bossk                 190    113  none       green              
60  Nute Gunray           191     90  none       mottled green      
61  Bail Prestor Organa   191     NA  black      tan                
62  San Hill              191     NA  none       grey               
63  Qui-Gon Jinn          193     89  brown      fair               
64  Dooku                 193     80  white      fair               
65  Wat Tambor            193     48  none       green, grey        
66  Jar Jar Binks         196     66  none       orange             
67  Kit Fisto             196     87  none       green              
68  Mas Amedda            196     NA  none       blue               
69  Ki-Adi-Mundi          198     82  white      pale               
70  Dexter Jettster       198    102  none       brown              
71  IG-88                 200    140  none       metal              
72  Darth Vader           202    136  none       white              
73  Rugor Nass            206     NA  none       green              
74  Tion Medon            206     80  none       grey               
75  Taun We               213     NA  none       grey               
76  Grievous              216    159  none       brown, white       
77  Roos Tarpals          224     82  none       grey               
78  Chewbacca             228    112  brown      unknown            
79  Lama Su               229     88  none       grey               
80  Tarfful               234    136  brown      brown              
81  Yarael Poof           264     NA  none       white              
82  Arvel Crynyd           NA     NA  brown      fair               
83  Finn                   NA     NA  black      dark               
84

In [27]:
arrange(starwars, height) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yoda         66    17 white      green      brown            896 male  mascu~
 2 Ratts T~     79    15 none       grey, blue unknown           NA male  mascu~
 3 Wicket ~     88    20 brown      brown      brown              8 male  mascu~
 4 Dud Bolt     94    45 none       blue, grey yellow            NA male  mascu~
 5 R2-D2        96    32 NA         white, bl~ red               33 none  mascu~
 6 R4-P17       96    NA none       silver, r~ red, blue         NA none  femin~
 7 R5-D4        97    32 NA         white, red red               NA none  mascu~
 8 Sebulba     112    40 none       grey, red  orange            NA male  mascu~
 9 Gasgano     122    NA none       white, bl~ black             NA male  mascu~
10 Watto       137    NA black      blue, grey yellow            NA male  mascu~
# i 77 m

The `arrange()` function, like most `dplyr` verb functions, takes the dataframe (`starwars`) as its first argument, and the variable names (e.g., `height`) as subsequent arguments.

In height order, we can see that Yoda is the shortest character, at 66 cm, with podracer [Ratts Tyerell](https://starwars.fandom.com/wiki/Ratts_Tyerell) next shortest, at 79 cm.

We can also arrange the dataframe in *descending* order, with the *`desc()`* function.

In [28]:
arrange(starwars, desc(height))

name                  height mass hair_color skin_color         
1   Yarael Poof           264     NA  none       white              
2   Tarfful               234    136  brown      brown              
3   Lama Su               229     88  none       grey               
4   Chewbacca             228    112  brown      unknown            
5   Roos Tarpals          224     82  none       grey               
6   Grievous              216    159  none       brown, white       
7   Taun We               213     NA  none       grey               
8   Rugor Nass            206     NA  none       green              
9   Tion Medon            206     80  none       grey               
10  Darth Vader           202    136  none       white              
11  IG-88                 200    140  none       metal              
12  Ki-Adi-Mundi          198     82  white      pale               
13  Dexter Jettster       198    102  none       brown              
14  Jar Jar Binks         196     66  none       orange             
15  Kit Fisto             196     87  none       green              
16  Mas Amedda            196     NA  none       blue               
17  Qui-Gon Jinn          193     89  brown      fair               
18  Dooku                 193     80  white      fair               
19  Wat Tambor            193     48  none       green, grey        
20  Nute Gunray           191     90  none       mottled green      
21  Bail Prestor Organa   191     NA  black      tan                
22  San Hill              191     NA  none       grey               
23  Bossk                 190    113  none       green              
24  Anakin Skywalker      188     84  blond      fair               
25  Mace Windu            188     84  none       dark               
26  Saesee Tiin           188     NA  none       pale               
27  Plo Koon              188     80  none       orange             
28  Raymus Antilles       188     79  brown      light              
29  Gregar Typho          185     85  black      dark               
30  Adi Gallia            184     50  none       dark               
... ...                   ...    ...  ...        ...                
58  Luminara Unduli       170    56.2 black      yellow             
59  Zam Wesell            168    55.0 blonde     fair, green, yellow
60  C-3PO                 167    75.0 NA         gold               
61  Jocasta Nu            167      NA white      fair               
62  Barriss Offee         166    50.0 black      yellow             
63  Beru Whitesun lars    165    75.0 brown      light              
64  Dormé                 165      NA brown      light              
65  Padmé Amidala         165    45.0 brown      light              
66  Shmi Skywalker        163      NA black      fair               
67  Ben Quadinaros        163    65.0 none       grey, green, yellow
68  Nien Nunb             160    68.0 none       grey               
69  Cordé                 157      NA brown      light              
70  Leia Organa           150    49.0 brown      light              
71  Mon Mothma            150      NA auburn     fair               
72  Watto                 137      NA black      blue, grey         
73  Gasgano               122      NA none       white, blue        
74  Sebulba               112    40.0 none       grey, red          
75  R5-D4                  97    32.0 NA         white, red         
76  R2-D2                  96    32.0 NA         white, blue        
77  R4-P17                 96      NA none       silver, red        
78  Dud Bolt               94    45.0 none       blue, grey         
79  Wicket Systri Warrick  88    20.0 brown      brown              
80  Ratts Tyerell          79    15.0 none       grey, blue         
81  Yoda                   66    17.0 white      green              
82  Arvel Crynyd           NA      NA brown      fair               
83  Finn                   NA      NA black      dark               
84

In [29]:
arrange(starwars, desc(height)) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yarael ~    264    NA none       white      yellow          NA   male  mascu~
 2 Tarfful     234   136 brown      brown      blue            NA   male  mascu~
 3 Lama Su     229    88 none       grey       black           NA   male  mascu~
 4 Chewbac~    228   112 brown      unknown    blue           200   male  mascu~
 5 Roos Ta~    224    82 none       grey       orange          NA   male  mascu~
 6 Grievous    216   159 none       brown, wh~ green, y~       NA   male  mascu~
 7 Taun We     213    NA none       grey       black           NA   fema~ femin~
 8 Rugor N~    206    NA none       green      orange          NA   male  mascu~
 9 Tion Me~    206    80 none       grey       black           NA   male  mascu~
10 Darth V~    202   136 none       white      yellow          41.9 male  mascu~
# i 77 m

This shows us that long-necked Jedi, [Yarael Poof](https://starwars.fandom.com/wiki/Yarael_Poof), is the tallest character, at 264 cm.

As well as sorting by numeric variables, we can also sort by character variables. For instance, we can sort by hair colour. If we do this we will sort hair alphabetically, from "auburn" to "white".

In [30]:
arrange(starwars, hair_color)

name                  height mass  hair_color    skin_color         
1   Mon Mothma            150       NA auburn        fair               
2   Wilhuff Tarkin        180       NA auburn, grey  fair               
3   Obi-Wan Kenobi        182     77.0 auburn, white fair               
4   Biggs Darklighter     183     84.0 black         light              
5   Boba Fett             183     78.2 black         fair               
6   Lando Calrissian      177     79.0 black         dark               
7   Watto                 137       NA black         blue, grey         
8   Quarsh Panaka         183       NA black         dark               
9   Shmi Skywalker        163       NA black         fair               
10  Eeth Koth             171       NA black         brown              
11  Gregar Typho          185     85.0 black         dark               
12  Luminara Unduli       170     56.2 black         yellow             
13  Barriss Offee         166     50.0 black         yellow             
14  Bail Prestor Organa   191       NA black         tan                
15  Jango Fett            183     79.0 black         tan                
16  Finn                   NA       NA black         dark               
17  Luke Skywalker        172     77.0 blond         fair               
18  Anakin Skywalker      188     84.0 blond         fair               
19  Finis Valorum         170       NA blond         fair               
20  Zam Wesell            168     55.0 blonde        fair, green, yellow
21  Leia Organa           150     49.0 brown         light              
22  Beru Whitesun lars    165     75.0 brown         light              
23  Chewbacca             228    112.0 brown         unknown            
24  Han Solo              180     80.0 brown         fair               
25  Wedge Antilles        170     77.0 brown         fair               
26  Jek Tono Porkins      180    110.0 brown         fair               
27  Arvel Crynyd           NA       NA brown         fair               
28  Wicket Systri Warrick  88     20.0 brown         brown              
29  Qui-Gon Jinn          193     89.0 brown         fair               
30  Ric Olié              183       NA brown         fair               
... ...                   ...    ...   ...           ...                
58  Mace Windu            188      84  none          dark               
59  Kit Fisto             196      87  none          green              
60  Adi Gallia            184      50  none          dark               
61  Saesee Tiin           188      NA  none          pale               
62  Yarael Poof           264      NA  none          white              
63  Plo Koon              188      80  none          orange             
64  Mas Amedda            196      NA  none          blue               
65  Poggle the Lesser     183      80  none          green              
66  Dexter Jettster       198     102  none          brown              
67  Lama Su               229      88  none          grey               
68  Taun We               213      NA  none          grey               
69  Ratts Tyerell          79      15  none          grey, blue         
70  R4-P17                 96      NA  none          silver, red        
71  Wat Tambor            193      48  none          green, grey        
72  San Hill              191      NA  none          grey               
73  Shaak Ti              178      57  none          red, blue, white   
74  Grievous              216     159  none          brown, white       
75  Sly Moore             178      48  none          pale               
76  Tion Medon            206      80  none          grey               
77  BB8                    NA      NA  none          none               
78  Captain Phasma         NA      NA  unknown       unknown            
79  Yoda                   66      17  white         green              
80  Ki-Adi-Mundi          198      82  white         pale     

In [31]:
arrange(starwars, hair_color) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Mon Mot~    150  NA   auburn     fair       blue            48   fema~ femin~
 2 Wilhuff~    180  NA   auburn, g~ fair       blue            64   male  mascu~
 3 Obi-Wan~    182  77   auburn, w~ fair       blue-gray       57   male  mascu~
 4 Biggs D~    183  84   black      light      brown           24   male  mascu~
 5 Boba Fe~    183  78.2 black      fair       brown           31.5 male  mascu~
 6 Lando C~    177  79   black      dark       brown           31   male  mascu~
 7 Watto       137  NA   black      blue, grey yellow          NA   male  mascu~
 8 Quarsh ~    183  NA   black      dark       brown           62   NA    NA    
 9 Shmi Sk~    163  NA   black      fair       brown           72   fema~ femin~
10 Eeth Ko~    171  NA   black      brown      brown           NA   male  mascu~
# i 77 m

Finally, we can sort by multiple variables at once. For instance, we may want to sort by hair colour alphabetically, and then by height descendingly:

In [32]:
arrange(starwars, hair_color, desc(height))

name                  height mass  hair_color    skin_color         
1   Mon Mothma            150       NA auburn        fair               
2   Wilhuff Tarkin        180       NA auburn, grey  fair               
3   Obi-Wan Kenobi        182     77.0 auburn, white fair               
4   Bail Prestor Organa   191       NA black         tan                
5   Gregar Typho          185     85.0 black         dark               
6   Biggs Darklighter     183     84.0 black         light              
7   Boba Fett             183     78.2 black         fair               
8   Quarsh Panaka         183       NA black         dark               
9   Jango Fett            183     79.0 black         tan                
10  Lando Calrissian      177     79.0 black         dark               
11  Eeth Koth             171       NA black         brown              
12  Luminara Unduli       170     56.2 black         yellow             
13  Barriss Offee         166     50.0 black         yellow             
14  Shmi Skywalker        163       NA black         fair               
15  Watto                 137       NA black         blue, grey         
16  Finn                   NA       NA black         dark               
17  Anakin Skywalker      188     84.0 blond         fair               
18  Luke Skywalker        172     77.0 blond         fair               
19  Finis Valorum         170       NA blond         fair               
20  Zam Wesell            168     55.0 blonde        fair, green, yellow
21  Tarfful               234    136.0 brown         brown              
22  Chewbacca             228    112.0 brown         unknown            
23  Qui-Gon Jinn          193     89.0 brown         fair               
24  Raymus Antilles       188     79.0 brown         light              
25  Ric Olié              183       NA brown         fair               
26  Cliegg Lars           183       NA brown         fair               
27  Han Solo              180     80.0 brown         fair               
28  Jek Tono Porkins      180    110.0 brown         fair               
29  Wedge Antilles        170     77.0 brown         fair               
30  Beru Whitesun lars    165     75.0 brown         light              
... ...                   ...    ...   ...           ...                
58  Mace Windu            188      84  none          dark               
59  Saesee Tiin           188      NA  none          pale               
60  Plo Koon              188      80  none          orange             
61  Adi Gallia            184      50  none          dark               
62  Poggle the Lesser     183      80  none          green              
63  Ackbar                180      83  none          brown mottle       
64  Bib Fortuna           180      NA  none          pale               
65  Ayla Secura           178      55  none          blue               
66  Shaak Ti              178      57  none          red, blue, white   
67  Sly Moore             178      48  none          pale               
68  Lobot                 175      79  none          light              
69  Darth Maul            175      80  none          red                
70  Ben Quadinaros        163      65  none          grey, green, yellow
71  Nien Nunb             160      68  none          grey               
72  Gasgano               122      NA  none          white, blue        
73  Sebulba               112      40  none          grey, red          
74  R4-P17                 96      NA  none          silver, red        
75  Dud Bolt               94      45  none          blue, grey         
76  Ratts Tyerell          79      15  none          grey, blue         
77  BB8                    NA      NA  none          none               
78  Captain Phasma         NA      NA  unknown       unknown            
79  Ki-Adi-Mundi          198      82  white         pale               
80  Dooku                 193      80  white         fair     

In [33]:
arrange(starwars, hair_color, desc(height)) |> print()

# A tibble: 87 x 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Mon Mot~    150  NA   auburn     fair       blue            48   fema~ femin~
 2 Wilhuff~    180  NA   auburn, g~ fair       blue            64   male  mascu~
 3 Obi-Wan~    182  77   auburn, w~ fair       blue-gray       57   male  mascu~
 4 Bail Pr~    191  NA   black      tan        brown           67   male  mascu~
 5 Gregar ~    185  85   black      dark       brown           NA   male  mascu~
 6 Biggs D~    183  84   black      light      brown           24   male  mascu~
 7 Boba Fe~    183  78.2 black      fair       brown           31.5 male  mascu~
 8 Quarsh ~    183  NA   black      dark       brown           62   NA    NA    
 9 Jango F~    183  79   black      tan        brown           66   male  mascu~
10 Lando C~    177  79   black      dark       brown           31   male  mascu~
# i 77 m

### Check your Knowledge!

Try coming up with code to solve the following:

1. Sort descendingly by mass, such that the most massive character comes first.

2. Sort by gender, eye colour, and then height. Which character is the first observation? Why?